core/src/main/java/io/parsingdata/metal/expression/value/reference/Self.java

```diff
@@ -16,22 +16,23 @@
 
 package io.parsingdata.metal.expression.value.reference;
 
+import java.util.Optional;
+
 import io.parsingdata.metal.Util;
-import io.parsingdata.metal.data.ImmutableList;
 import io.parsingdata.metal.data.ParseState;
 import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.expression.value.SingleValueExpression;
 import io.parsingdata.metal.expression.value.Value;
-import io.parsingdata.metal.expression.value.ValueExpression;
 
 /**
- * A {@link ValueExpression} that represents the {@link Value} most recently
+ * A {@link SingleValueExpression} that represents the {@link Value} most recently
  * added to the parse state.
  */
-public class Self implements ValueExpression {
+public class Self implements SingleValueExpression {
 
     @Override
-    public ImmutableList<Value> eval(final ParseState parseState, final Encoding encoding) {
-        return parseState.order.current().map(ImmutableList::<Value>create).orElseGet(ImmutableList::new);
+    public Optional<Value> evalSingle(final ParseState parseState, final Encoding encoding) {
+        return parseState.order.current().flatMap(Optional::of);
```

Comment:
Ah, here is the upcast :wink:

Reply:
Not `.map(identity())`? :stuck_out_tongue_closed_eyes:

Reply:
@ccreeten: Oh I actually like that one better...